### DISCUSSION TITLE
How can I resample image into another image to make them have common affine?

### DISCUSSION THREAD

**Discussion by karwojan** [Project-MONAI/MONAI#3842] [Q&A] [answered]
I'm preparing some easy flow of transformations to load MR images, change spacing to 1x1x1 and get random crop. I'm loading image and segmentation using single compose dict transforms (with two keys). The problem is that images and segmentations have a little bit different affines (but are in the same place in the physical/reference space). Before I go with any other transform (fix spacing, augment, etc.) I want to these two images have common afine - i.e. I want to resample segmentation into image space. How can I do that? Is it possible to include it within transforms chain? 

In SimpleITK it I use sitk.Resample(image, referenceImage).

**Comment by wyli** [comment, marked as answer]
great question, there is a `ResampleToMatch` transform: https://github.com/Project-MONAI/MONAI/blob/2cfb896d99507147006569a7ec3cc1c068b6b096/tests/test_resample_to_matchd.py#L38-L43

https://github.com/Project-MONAI/MONAI/issues/3823

**Reply by karwojan** [reply]
Thank you for response. This transform is exactly what I need. I've tried to use it but it changes shape of my input image - looks like "dim" value in "meta_dict" of my image (which is used in ResampleToMatch) differs from raw image.shape. This just after LoadImage transform:
```
>>> data["image_meta_dict"]["dim"]
array([  3, 480, 512, 176,   1,   1,   1,   1], dtype=int16)
>>> data["image"].shape
(176, 480, 512)
```
Is this proper behaviour? If not what could be the reason?

**Reply by wyli** [reply]
thanks, looks like your input is a 3d spatial image without a channel dimension, perhaps you can try  `LoadImaged("image", channel_dim=None)` or
```py
LoadImaged("image", ensure_channel_first=True)
```
with the latest version of monai installed from github https://docs.monai.io/en/latest/installation.html#from-github

if this doesn't work, would be great if you could include more details for debugging.

**Reply by karwojan** [reply]
Ok, I found the problem. I've prepared some wrapper for NibabelReader and there was a mistake in positional arguments - I accidentially used as_closest_canonical=True which changed shape of my image. But in fact, when you use this flag then ResampleToMatchd works following way:
```
>>> data = {"image": "/tmp/image.nii.gz", "labels": "/tmp/labels.mgz"}
>>> data = LoadImaged(["image", "labels"], reader=NibabelReader(as_closest_canonical=True))(data)
>>> data = AddChanneld(["image", "labels"])(data)
>>> data["image"].shape
(1, 176, 480, 512)
>>> data["labels"].shape
(1, 176, 480, 512)
>>> data = ResampleToMatchd(["labels"], "image_meta_dict", mode="nearest")(data)
>>> data["image"].shape
(1, 176, 480, 512)
>>> data["labels"].shape
(1, 480, 512, 176)
```

**Reply by karwojan** [reply]
By the way, I wonder if there is one more issue. In some cases, ResampleToMatchd changes affine to single value:
```
>>> data = {"image": "/tmp/image.nii.gz", "labels": "/tmp/labels.mgz"}
>>> data = LoadImaged(["image", "labels"], reader=NibabelReader(as_closest_canonical=False))(data)
>>> data = AddChanneld(["image", "labels"])(data)
>>> data["image_meta_dict"]["affine"]
array([[-1.70929562e-02,  6.90474268e-03,  8.99303675e-01,
        -7.69320374e+01],
       [-4.62404788e-01,  7.47344792e-02, -3.45145576e-02,
         1.12282265e+02],
       [-7.49414563e-02, -4.62702543e-01,  7.84528535e-03,
         1.47994492e+02],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         1.00000000e+00]])
>>> data = ResampleToMatchd(["image"], "labels_meta_dict", mode="nearest")(data)
>>> data["image_meta_dict"]["affine"]
array([[-0.01709295]], dtype=float32)
```
It does not happen if I resample from "labels" to "image".

**Reply by wyli** [reply]
yeah, I put a patch here https://github.com/Project-MONAI/MONAI/pull/3847/files, might address this issue, I'll look into it  cc @rijobro

**Reply by rijobro** [reply]
thanks @karwojan for flagging these bugs. The `ResampleToMatch` transform is new, so any help you can give in perfecting it would be greatly appreciated. Thanks for the PR @wyli!
